### PR TITLE
Outlink tracking when DOM modified

### LINKED ISSUE #15780: Outlink and download tracking not working for links added after DOM ready (which many are these days eg angular, vue.js, react, ... SPA).
Hi there,

We are using Matomo on a VueJS project but we have trouble tracking outlinks.

When a user navigates from our home page to an other page inside the app and then clicks on a link that redirects him outside the app, the link does not appear as an outlink on the Matomo dashboard. However, when the user clicks on an external link directly on the home page, the outlink appears correctly on the Matomo dasboard.

You can find a very simple vue project where you can reproduce the issue: https://github.com/thibautguedou3/matomo-vue-outlink

I can reproduce the issue both on Chrome and Safari

Platform: Mac OS

Could it be a configuration issue in the matomo script ?
```html
<script type="text/javascript">
      var _paq = window._paq || [];
      /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
      _paq.push(["trackPageView"]);
      _paq.push(["enableLinkTracking"]);
      (function() {
        var u = "MATOMO_URL";
        _paq.push(["setTrackerUrl", u + "matomo.php"]);
        _paq.push(["setSiteId", "SITE_ID"]);
        var d = document,
          g = d.createElement("script"),
          s = d.getElementsByTagName("script")[0];
        g.type = "text/javascript";
        g.async = true;
        g.defer = true;
        g.src = "//cdn.matomo.cloud/MATOMO_URL/matomo.js";
        s.parentNode.insertBefore(g, s);
      })();
    </script>
```

### PR DESCRIPTION
**Pull request by flamisz**
### Description:

fixes #15780

### Review

* [ ] Functional review done
* [ ] Potential edge cases thought about (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] Usability review done (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] Security review done [see checklist](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] Code review done
* [ ] Tests were added if useful/possible
* [ ] Reviewed for breaking changes
* [ ] Developer changelog updated if needed
* [ ] Documentation added if needed
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by flamisz** [comment]
build js

**Comment by flamisz** [comment]
build js

**Comment by diosmosis** [comment]
I see how this works, using the root element to always handle clicks (and it's definitely simpler), but I think there's potentially an issue if event bubbling is cancelled (ie, via stopPropagation()) somewhere along the way. In that case we might not track any outlinks. Is this something we want to handle @tsteur?

**Comment by tsteur** [comment]
@diosmosis that's something we can't prevent but we set the third parameter in addEventListener to `true` so we should be receiving the event as they are usually not cancelled in the `useCapture` mode.

From https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

> A Boolean indicating whether events of this type will be dispatched to the registered listener before being dispatched to any EventTarget beneath it in the DOM tree

it basically starts in the html/body elements and then goes down before it would bubble up again AFAIK. And usually someone might stop propagation only at a later stage (they have default useCapture=false). It works quite well eg in TagManager etc and it shouldn't be a problem

File: js/piwik.js [hunk 1]
```diff
@@ -4627,32 +4652,20 @@ if (typeof window.Matomo !== 'object') {
...
-                                // we make sure to setup link only once for each tracker
-                                linkElement.matomoTrackers.push(trackerInstance);
-                                addClickListener(linkElement, enable);
-                            }
-                        }
-                    }
+                if (enable) {
+                    addEventListener(documentAlias.body, 'mouseup', clickHandler(enable), true);
+                    addEventListener(documentAlias.body, 'mousedown', clickHandler(enable), true);
+                    addEventListener(documentAlias.body, 'contextmenu', clickHandler(enable), true);
```
Comment (diosmosis) [inline js/piwik.js]: This is basically the same code as `addClickListener()` except w/ useCapture = true, right? If so, adding a useCapture parameter to addClickListener, and then calling it here would mean less code, would that be better?

**Comment by diosmosis** [comment]
Tested locally, works well. Created https://github.com/matomo-org/test-examples/pull/7 for testing.

Also left one small comment, otherwise looks good.

**Comment by tsteur** [comment]
fyi because the 4.3 release is soon and the RC is already out we might want to merge that one only after the 4.3 release just in case anything is breaking etc.  And then we could directly update the SPA guide etc to mention the method no longer needs to be called etc

**Comment by diosmosis** [comment]
@tsteur should we change the milestone?

**Comment by tsteur** [comment]
Changed it for now 👍  Hoping it won't be later renamed to 4.5.0 after releasing 4.3. Then we would change the milestone again.

I could create a 4.3.1 milestone but there we'd mostly want fixes and not changes in there.

File: js/piwik.js (hunk 1 above)
Comment (flamisz) [reply js/piwik.js]: yeah, I didn't realize it, after all the modifications basically they are the same. I'll do it

Comment (flamisz) [reply js/piwik.js]: updated

**Comment by flamisz** [comment]
build js

**Pull request merged**